### PR TITLE
Add support for TF export

### PR DESCRIPTION
**Pull request by AyushExel**
Usage:
```
python export.py --weights yolov5s-seg.pt --include tflite
```
Output
Screenshot from 2022-08-30 13-20-58

We'll need to wait for the official models to complete training before we can run benchmark on exported model

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@AyushExel nice! I think TFProto needs a call() method though no?

**Comment by glenn-jocher** [comment]
@AyushExel also you should build and run a segmentation model following the detection example below to verify it works, i.e. from tf.py run() function:

```python
    # TensorFlow model
    im = tf.zeros((batch_size, *imgsz, 3))  # BHWC image
    tf_model = TFModel(cfg='yolov5-seg.yaml', model=model, nc=model.nc, imgsz=imgsz)
    _ = tf_model.predict(im)  # inference
```

EDIT: TFSegment() forward method needs to be renamed to 'call' in TF terminology

**Comment by AyushExel** [comment]
@glenn-jocher ok.
Channel is at last in tf format right. So the input of proto layer should be `1x80x80x64` instead of `1x64x80x80` but it still complains about dimensions.
Screenshot 2022-08-30 at 4 49 07 PM
Screenshot 2022-08-30 at 4 49 41 PM
I think there's something messing up the weight transfer, I'm jumping into this rabbit hole

**Comment by AyushExel** [comment]
@glenn-jocher can you tag the tf guy here? I can't recall his username.. It's something zldroid?

**Comment by glenn-jocher** [comment]
@zldrobit can you take a look at this PR please? We are adding Segmentation model support to YOLOv5 in https://github.com/ultralytics/yolov5/pull/9052, and we need to add two additional modules to tf.py which are a Segment() head and a Proto() module for masks. Segment() is a little tricky as it inherits from Detect() and uses the Detect.forward method in addition to Proto.forward.

https://github.com/AyushExel/yolov5/blob/3b25f3c54d96cf7c317a4e9410027726d15bf892/models/yolo.py#L92-L107

https://github.com/AyushExel/yolov5/blob/3b25f3c54d96cf7c317a4e9410027726d15bf892/models/common.py#L764-L775

**Comment by zldrobit** [comment]
@glenn-jocher @AyushExel I am trying to reproduce the error. However, I couldn't find the segmentation model (`yolov5s-seg.pt`). If it's possible, plz share the segmentation model for debugging.

EDIT: By comparing https://github.com/AyushExel/yolov5/blob/3b25f3c54d96cf7c317a4e9410027726d15bf892/models/yolo.py#L92-L107
https://github.com/AyushExel/yolov5/blob/3b25f3c54d96cf7c317a4e9410027726d15bf892/models/common.py#L764-L775
with
https://github.com/AyushExel/yolov5/blob/d2af8e1337b77a870c04c7542c56c0aea3dabefa/models/tf.py#L322-L347
, I cannot find any mismatch numbers.

**Comment by zldrobit** [comment]
@AyushExel 

>Channel is at last in tf format right. So the input of proto layer should be 1x80x80x64 instead of 1x64x80x80 but it still complains about dimensions.





>I think there's something messing up the weight transfer, I'm jumping into this rabbit hole

From the second image, I recognized that 36864 = 147456 / 4 = 3 x 3 x 64 x 256 / 4. Maybe this is helpful to locate the problem.

**Comment by glenn-jocher** [comment]
@zldrobit thanks for taking a look! The new v6.3 segmentation models are temporarily in the v6.2 assets. They are just finishing training now. I've uploaded yolvo5s-seg.pt here:
https://github.com/ultralytics/yolov5/releases/download/v6.2/yolov5s-seg.pt

**Comment by zldrobit** [comment]
@glenn-jocher glad to help! I sent [a PR](https://github.com/AyushExel/yolov5/pull/25) to address the TF export problem of segmentation model. After that, I tried to run `segment/predict.py` with the TFLite model but failed. TFLite inference in `DetectMultiBackend()` assumes TF/TFLite models have only one output, and it's conflicted with the segmentation model of more than one output. IMHO, running YOLOv5 TF segment model may take some more development and refactor efforts.

**Comment by AyushExel** [comment]
@zldrobit thanks! I think your last PR that I just merged solves this problem?

**Comment by zldrobit** [comment]
@AyushExel my pleasure! Yes, I could confirm after the PR, the `instance_seg_tf` branch could export a YOLOv5 TF/TFLite segmentation model.

**Comment by glenn-jocher** [comment]
@zldrobit awesome thanks for the help! 

@AyushExel allright let's merge this into instance_seg and I can debug a bit there.

**Comment by glenn-jocher** [comment]
@zldrobit yes it looks like DetectMultiBackend is expecting all models to output a single np.array/torch.tensor, but we have lists/tuples used for Segmentation, so we need some conditional logic in DetectMultiBackend to handle SegmentationModels specially, and/or we need to convert ClassificationModel and DetectionModel to output list/tuple also of length 1 to unify all the YOLOv5 models.
